code/iaas/ha/src/test/java/io/cattle/platform/ha/monitor/impl/PingInstancesMonitorImplTest.java

```diff
@@ -0,0 +1,274 @@
+            }
+        };
+        knownInstances = new HashMap<String, Instance>();
+        needsSynced = new HashMap<String, ReportedInstance>();
+        syncActions = new HashMap<String, String>();
+        needsHaRestart = new HashSet<String>();
+        reportedInstances = new ReportedInstances();
+    }
+
+    String setupSync(String key, boolean byExternalId, boolean sysCon, String knownInstanceState, 
+            String reportedInstanceState) {
+        return setupSync(key, byExternalId, sysCon, null, knownInstanceState, reportedInstanceState);
+    }
+    String setupSync(String key, boolean byExternalId, boolean sysCon, String instanceTriggeredStop, String knownInstanceState, 
+            String reportedInstanceState) {
+        key = fullKey(key, byExternalId, sysCon);
+        String uuid = key + "-uuid";
+        String externalId = key + "-extid";
+
+        String instaceEid = byExternalId ? externalId : null;
+        String instanceSysCon = sysCon ? "ImASysCon" : null;
+        Date instanceRemoved = REMOVED.equals(knownInstanceState) ? new Date() : null;
+        String reportedUuid = !byExternalId ? uuid : "garbage-" + key;
+
+        if (knownInstanceState != null)
+            addKnownInstance(uuid, instaceEid, knownInstanceState, instanceSysCon, instanceRemoved, instanceTriggeredStop);
+        if (reportedInstanceState != null)
+            addReportedInstance(reportedUuid, externalId, reportedInstanceState, sysCon);
+
+        return externalId;
+    }
+
+    String fullKey(String key, boolean byExternalId, boolean sysCon) {
+        StringBuilder fullKey = new StringBuilder();
+        if (sysCon)
+            fullKey.append("syscon-");
+        else
+            fullKey.append("user-");
+
+        fullKey.append(key);
+
+        if (byExternalId)
+            fullKey.append("-byext");
+        else
+            fullKey.append("-byuuid");
+
+        return fullKey.toString();
+    }
+
+    String uuid(String key, boolean byExternalId, boolean sysCon) {
+        String fullKey = fullKey(key, byExternalId, sysCon);
+        return fullKey + "-uuid";
+    }
+
+    @Test
+    public void testDetermineSyncActions() {
+        // container running in rancher, stopped on host
+        String externalIdA = setupSync("rancher-running-host-stopped", false, false, STATE_RUNNING, STATE_STOPPED);
+        String externalIdC = setupSync("rancher-running-host-stopped", true, false, STATE_RUNNING, STATE_STOPPED);
+        String externalIdB = setupSync("rancher-running-host-stopped", false, true, STATE_RUNNING, STATE_STOPPED);
+        String externalIdC1 = setupSync("rancher-running-host-stopped", true, true, STATE_RUNNING, STATE_STOPPED);
+        String externalIdC2 = setupSync("rancher-running-host-stopped-triggered", false, false, "restart", STATE_RUNNING, STATE_STOPPED);
+        String externalIdC3 = setupSync("rancher-running-host-stopped-triggered", true, false, "restart", STATE_RUNNING, STATE_STOPPED);
```

Comment:
@alena1108 FYI, these two "rancher-running-host-stopped-triggered" cases are the tests i added to test the instanceTriggeredStop + not-a-system-container logic change.